<h2 style="text-align:center"><strong>Business Problem</strong></h2>

#### The e-commerce company needs a comprehensive analysis of its data to uncover insights into customer demographics and purchasing behavior, product and category performance, revenue growth and discount impact, supplier contributions, and shipping efficiency. The challenge is to integrate these perspectives into a single view that helps management identify high-value customers, top-performing products, reliable suppliers, and efficient shippers, enabling better pricing, marketing, and operational strategies to drive profitability and retention.


## Database Connection & Query Execution Utility

In [ ]:
import pandas as pd
import sqlalchemy
from sqlalchemy.engine import URL

# Database connection setup (general form)
connection_url = URL.create(
    drivername="mysql+pymysql",   # database + driver
    username="your_username",     # replace with your DB username
    password="your_password",     # replace with your DB password
    host="your_host",             # e.g., "localhost" or server IP
    port=3306,                    # default MySQL port
    database="your_database"      # replace with your DB name
)

# Create the SQLAlchemy engine
engine = sqlalchemy.create_engine(connection_url) 

In [255]:
def run_query(query, title=None):
    if title:
        print(f"\n--- {title} ---\n")
    return pd.read_sql(query, engine)

---

## Displaying data from all the tables 

#### Customers Table

In [256]:
run_query("""SELECT * FROM customers LIMIT 10;""")

CustomerID FirstName   LastName Date_of_Birth         City          State  \
0       57081     James      Smith    1987-03-26     New York       New York   
1       57082    Robert  Downey Jr    1973-05-24     New York       New York   
2       57083      John   Williams    1990-04-14      Chicago       Illinois   
3       57084   Michael    Johnson    1953-03-25     Brisbane     Queensland   
4       57085     Steve   Williams    1971-04-24       Bremen         Bremen   
5       57086     David    Beckham    1965-10-18      Villach      Carinthia   
6       57087   Richard      Brown    1958-12-16  San Antonio          Texas   
7       57088    Joseph      James    1967-08-27    Amsterdam  North Holland   
8       57089    Thomas      Jones    1950-04-30       Dallas          Texas   
9       57090   Charles       King    1993-06-24       Warsaw       Masovian   

         Country  PostalCode       Phone                       Email  \
0  United States      280862  9638483934       James.Smith@gmail.com   
1  United States      376573  6588282115  Robert.Downey Jr@gmail.com   
2  United States      485629  7641021429     John.Williams@gmail.com   
3      Australia      260866  7354232181   Michael.Johnson@gmail.com   
4        Germany      740338  6285552036    Steve.Williams@gmail.com   
5        Austria      621492  7756781677     David.Beckham@gmail.com   
6  United States      110945  8420993031     Richard.Brown@gmail.com   
7    Netherlands      186896  6691466381      Joseph.James@gmail.com   
8  United States      174080  8024463594      Thomas.Jones@gmail.com   
9         Poland      589838  8498546902      Charles.King@gmail.com   

  DateEntered  
0  2020-01-02  
1  2020-01-06  
2  2020-01-11  
3  2020-01-16  
4  2020-01-17  
5  2020-01-20  
6  2020-01-20  
7  2020-01-25  
8  2020-01-26  
9  2020-01-27

#### Orders Table

In [257]:
run_query("""SELECT * FROM orders LIMIT 10;""")

OrderID  CustomerID  PaymentID   OrderDate  ShipperID    ShipDate  \
0  7655500       57083          2  2020-01-12          7  2020-01-13   
1  7655501       57086          3  2020-01-20          2  2020-01-24   
2  7655502       57086          4  2020-02-06          7  2020-02-11   
3  7655503       57088          4  2020-02-09          1  2020-02-13   
4  7655504       57090          4  2020-02-11          3  2020-02-15   
5  7655505       57094          4  2020-02-15          1  2020-02-24   
6  7655506       57092          4  2020-02-20          8  2020-02-23   
7  7655507       57095          6  2020-02-23          8  2020-03-04   
8  7655508       57105          4  2020-02-26          5  2020-02-28   
9  7655509       57095          5  2020-02-26          7  2020-02-27   

  DeliveryDate  Total_order_amount  
0   2020-01-19               25112  
1   2020-01-27               22453  
2   2020-02-21               13293  
3   2020-02-26               16063  
4   2020-02-20               15193  
5   2020-02-27               13581  
6   2020-02-27               74120  
7   2020-03-11               12092  
8   2020-03-08               48258  
9   2020-03-08                7509

#### Products Table

In [258]:
run_query("""SELECT * FROM products LIMIT 10;""")

ProductID                                            Product  Category_ID  \
0          1        Original Disinfectant Toilet Cleaner Liquid         5001   
1          2                         Surface Disinfectant Spray         5001   
2          3  Harpic Disinfectant Toilet Cleaner Original200...         5001   
3          4  Harpic Toilet Cleaner Liquid - Original 1 L + ...         5001   
4          5       Disinfectant Bathroom Cleaner Liquid - Lemon         5001   
5          6  Super Saver Pack Toilet Cleaner Original, 500m...         5001   
6          7     Original 1 ltr + Bathroom Cleaner Lemon 500 ml         5001   
7          8  Disinfectant Toilet Cleaner Liquid - 1 L + Bat...         5001   
8          9  Harpic Bathroom Cleaner - Lemon 1 L + Harpic T...         5001   
9         10  Lizol Floor Cleaner, Floral - 2L + Harpic Toil...         5001   

           Sub_Category     Brand  Sale_Price  Market_Price  \
0  All Purpose Cleaners    Harpic         489           534   
1  All Purpose Cleaners    Savlon         257           318   
2  All Purpose Cleaners  bb Combo          74            76   
3  All Purpose Cleaners  bb Combo         462           558   
4  All Purpose Cleaners    Harpic         299           362   
5  All Purpose Cleaners    Harpic         162           184   
6  All Purpose Cleaners    Harpic         247           273   
7  All Purpose Cleaners    Harpic         247           273   
8  All Purpose Cleaners    Harpic         318           359   
9  All Purpose Cleaners  bb Combo         483           540   

                            Type  
0                Toilet Cleaners  
1  Disinfectant Spray & Cleaners  
2                Toilet Cleaners  
3                Toilet Cleaners  
4                Toilet Cleaners  
5                Toilet Cleaners  
6                Toilet Cleaners  
7                Toilet Cleaners  
8                Toilet Cleaners  
9         Floor & Other Cleaners

#### Order Details Table

In [259]:
run_query("""SELECT * FROM orderdetails LIMIT 10;""")

OrderDetailID  OrderID  ProductID  Quantity  SupplierID
0              1  7655500      14955         2           3
1              2  7655500      19946         4           2
2              3  7655500       6299        18           2
3              4  7655500      12711         9           4
4              5  7655500      12743         1           3
5              6  7655500      17791         2           4
6              7  7655500       9268        17           4
7              8  7655500       8703        14           1
8              9  7655501      17335        14           3
9             10  7655501      13301        19           1

#### Category Table

In [260]:
run_query("""SELECT * FROM category;""")

CategoryID              CategoryName Active
0         5001      Cleaning & Household    Yes
1         5002    Kitchen, Garden & Pets    Yes
2         5003  Foodgrains, Oil & Masala    Yes
3         5004      Gourmet & World Food    Yes
4         5005                 Baby Care     No
5         5006    Snacks & Branded Foods    Yes
6         5007     Bakery, Cakes & Dairy    Yes
7         5008          Beauty & Hygiene    Yes
8         5009                 Beverages    Yes
9         5010         Eggs, Meat & Fish     No
10        5011       Fruits & Vegetables    Yes

####  Payments Table

In [261]:
run_query("""SELECT * FROM payments ;""")

PaymentID  PaymentType Allowed
0           1   Debit Card     Yes
1           2          POD     Yes
2           3       PayPal     Yes
3           4  Credit Card     Yes
4           5       Wallet     Yes
5           6  Net banking     Yes
6           7     PayLater      No
7           8     FoodCard      No
8           1   Debit Card     Yes
9           2          POD     Yes
10          3       PayPal     Yes
11          4  Credit Card     Yes
12          5       Wallet     Yes
13          6  Net banking     Yes
14          7     PayLater      No
15          8     FoodCard      No

#### Shippers Table

In [262]:
run_query("""SELECT * FROM shippers;""")

ShipperID                           CompanyName       Phone
0          1                             Blue Dart  9273377892
1          2                                  DTDC  9814107260
2          3                       Lufthansa Cargo  9356212514
3          4                                Fed Ex  7422454116
4          5                             Delhivery  8922763991
5          6  COSCO - China Ocean Shipping Company  8219297421
6          7                           Hapag-Lloyd  7891857833
7          8           ONE - Ocean Network Express  7400327522
8          9                        Evergreen Line  8021928034
9         10               Hyundai Merchant Marine  7464599377

#### Suppliers Table

In [263]:
run_query("""SELECT * FROM suppliers;""")

SupplierID                     CompanyName           City        State  \
0           1  Sunshine Electronics Pvt. Ltd.        Fremont   California   
1           2                   Tasha Apparel     Abu Dhabhi   Abu Dhabhi   
2           3                   Alpha imports         Austin        Texas   
3           4                    GreenDroShip      Bangalore    Karnataka   
4           5                       Megagoods        Kolkata  West Bengal   
5           6                    ASI Partners  New York City     New York   

   PostalCode Country       Phone                          Email  
0      841726     USA  8796300412          contactus@sunelec.com  
1      683008     UAE  8773744676  customercare@tashaapparel.com  
2      889037     USA  9872259654     customers@alphaimports.com  
3      416556   India  8245273210        gogreen@greendrop.co.in  
4      119658   India  8568274416          gobig@megagoods.co.in  
5      865066     USA  9363378404              contactus@asi.com

## Table Exploration

#### **Customers Table:** Stores customer information including personal details (name, date of birth), contact details (email, phone, address), and registration date. Useful for analyzing customer demographics, locations, and when they joined the platform.

#### **Orders Table:** Captures order-level details such as customer, payment method, order and shipping dates, and total order value. Provides the basis for analyzing order volumes, revenue trends, and delivery timelines.

#### **Products Table:** Contains product catalog information including category, sub-category, brand, sale price, market price, and type. Helps analyze product performance, pricing gaps, and category contributions.

#### **OrderDetails Table:** Breaks down each order into line items with product, quantity, and supplier mapping. Enables detailed sales analysis at the product and supplier level.

#### **Category Table:** Maintains product category definitions with active/inactive status. Supports grouping products into logical categories for performance and portfolio analysis.

#### **Payments Table:** Defines available payment methods and whether they are allowed. Useful for understanding customer payment preferences and payment channel restrictions.

#### **Shippers Table:** Stores shipping company information including names and contact details. Used to analyze shipping partners’ order volumes and delivery performance.

#### **Suppliers Table:** Contains supplier details such as company, location, and contact information. Supports analysis of supplier contributions to product availability and revenue.

---

## Customers
#### This section analyzes customer demographics, engagement, and purchasing behavior to provide insights into the company’s core user base. By understanding where customers are located, how recently they joined, how often they buy, and how much they spend, management can identify loyal buyers, active markets, and emerging opportunities. These insights are crucial for targeted marketing campaigns, customer retention programs, and personalized engagement strategies.

#### Customer details such as full name, email, and city are retrieved to build a foundation for customer segmentation and communication.

In [264]:
run_query("""
SELECT CONCAT(FirstName, ' ', LastName) AS Full_Name, Email, City FROM customers;""")


Full_Name                       Email       City
0         James Smith       James.Smith@gmail.com   New York
1    Robert Downey Jr  Robert.Downey Jr@gmail.com   New York
2       John Williams     John.Williams@gmail.com    Chicago
3     Michael Johnson   Michael.Johnson@gmail.com   Brisbane
4      Steve Williams    Steve.Williams@gmail.com     Bremen
..                ...                         ...        ...
520     Darragh Niall     Darragh.Niall@gmail.com    Hamburg
521     Dean Alistair     Dean.Alistair@gmail.com    Villach
522       Idris Colby       Idris.Colby@gmail.com     Dublin
523     Jesse Francis     Jesse.Francis@gmail.com       Rome
524    Malcolm Julian    Malcolm.Julian@gmail.com  Amsterdam

[525 rows x 3 columns]

#### Customer counts by country are calculated to evaluate geographic distribution and identify top-performing regions.

In [265]:
run_query("""
SELECT Country, COUNT(CustomerID) AS cnt 
FROM customers
GROUP BY Country 
ORDER BY cnt DESC;
""")

Country  cnt
0      United States   66
1              India   40
2           Portugal   37
3             France   35
4        Switzerland   33
5            Austria   32
6          Australia   30
7   Northern Ireland   30
8            Germany   29
9             Greece   28
10           Ireland   27
11           Belgium   27
12       Netherlands   24
13            Russia   24
14           Romania   20
15            Poland   19
16             Italy   14
17       New Zealand    5
18      South Africa    5

#### The most recent customers are listed based on registration date to track new acquisitions.

In [266]:
run_query("""
SELECT CustomerID, FirstName, DateEntered 
FROM customers
ORDER BY DateEntered ASC;
""")


CustomerID FirstName DateEntered
0         57081     James  2020-01-02
1         57082    Robert  2020-01-06
2         57083      John  2020-01-11
3         57084   Michael  2020-01-16
4         57085     Steve  2020-01-17
..          ...       ...         ...
520       57601   Darragh  2021-12-09
521       57602      Dean  2021-12-11
522       57603     Idris  2021-12-19
523       57604     Jesse  2021-12-22
524       57605   Malcolm  2021-12-28

[525 rows x 3 columns]

#### Customers who have placed at least one order are identified to measure active user engagement.

In [267]:
run_query("""
SELECT c.CustomerID, c.FirstName, COUNT(o.OrderID) AS cnt 
FROM customers AS c 
JOIN orders AS o ON c.CustomerID = o.CustomerID 
GROUP BY c.CustomerID, c.FirstName
HAVING COUNT(o.OrderID) >= 1
ORDER BY cnt DESC;
""")

CustomerID  FirstName  cnt
0         57334       Luca   19
1         57543        Rex   18
2         57125  Alexander   17
3         57213    Shirley   17
4         57383      Roman   17
..          ...        ...  ...
520       57542     Oakley    3
521       57565    Lincoln    3
522       57594      River    3
523       57506     Alfred    2
524       57165    Gabriel    1

[525 rows x 3 columns]

#### Total spending per customer is calculated to spot high-value buyers who contribute most to revenue.

In [268]:
run_query("""
SELECT c.CustomerID, 
       CONCAT(FirstName, ' ', LastName) AS Fullname, 
       SUM(od.Quantity * p.Sale_Price) AS total_amount_spent 
FROM customers AS c 
JOIN Orders AS o ON c.CustomerID = o.CustomerID 
JOIN OrderDetails AS od ON o.OrderID = od.OrderID 
JOIN Products AS p ON p.ProductID = od.ProductID 
GROUP BY c.CustomerID, Fullname 
ORDER BY total_amount_spent DESC;
""")


CustomerID              Fullname  total_amount_spent
0         57249  Jacqueline Fernandez            455604.0
1         57495          Toby Richard            425474.0
2         57232          Diane Morgan            410641.0
3         57574         Richard Corey            410011.0
4         57486           Cruz Duncan            408481.0
..          ...                   ...                 ...
520       57506        Alfred Bradley             35525.0
521       57565        Lincoln Arthur             34122.0
522       57261      Abigail Richards             22527.0
523       57442         Miller Thomas             20607.0
524       57501           Kacper Kian             11634.0

[525 rows x 3 columns]

#### Customers are ranked by order frequency to find repeat buyers and frequent shoppers.

In [269]:
run_query("""
SELECT c.CustomerID, 
       c.FirstName, 
       COUNT(o.OrderID) AS cnt, 
       ROW_NUMBER() OVER(ORDER BY COUNT(o.OrderID) DESC) AS rnk
FROM customers AS c 
JOIN orders AS o ON c.CustomerID = o.CustomerID 
GROUP BY c.CustomerID, c.FirstName;
""")

CustomerID  FirstName  cnt  rnk
0         57334       Luca   19    1
1         57543        Rex   18    2
2         57125  Alexander   17    3
3         57213    Shirley   17    4
4         57383      Roman   17    5
..          ...        ...  ...  ...
520       57542     Oakley    3  521
521       57565    Lincoln    3  522
522       57594      River    3  523
523       57506     Alfred    2  524
524       57165    Gabriel    1  525

[525 rows x 4 columns]

#### Customers placing consecutive orders within 7 days are highlighted to detect highly engaged users.

In [270]:
run_query("""
WITH OrderDates AS (
    SELECT c.CustomerID,
           o.OrderID,
           o.OrderDate,
           LAG(o.OrderDate) OVER (PARTITION BY c.CustomerID ORDER BY o.OrderDate) AS PreviousOrderDate
    FROM Orders AS o 
    JOIN Customers AS c ON o.CustomerID = c.CustomerID
)
SELECT CustomerID,
       OrderID,
       OrderDate,
       PreviousOrderDate,
       DATEDIFF(OrderDate, PreviousOrderDate) AS DaysBetweenOrders
FROM OrderDates
WHERE PreviousOrderDate IS NOT NULL
  AND DATEDIFF(OrderDate, PreviousOrderDate) <= 7
ORDER BY CustomerID, OrderDate;
""")

CustomerID  OrderID   OrderDate PreviousOrderDate  DaysBetweenOrders
0          57083  7655522  2020-03-20        2020-03-18                  2
1          57083  7656716  2021-03-30        2021-03-30                  0
2          57084  7656338  2021-01-26        2021-01-23                  3
3          57085  7656430  2021-02-14        2021-02-09                  5
4          57086  7656745  2021-04-03        2021-04-01                  2
...          ...      ...         ...               ...                ...
1211       57605  7660468  2021-12-28        2021-12-28                  0
1212       57605  7660493  2021-12-28        2021-12-28                  0
1213       57605  7660494  2021-12-28        2021-12-28                  0
1214       57605  7660496  2021-12-28        2021-12-28                  0
1215       57605  7660479  2021-12-28        2021-12-28                  0

[1216 rows x 5 columns]

#### High-Value Customers (Revenue Ranking)

In [271]:
run_query("""
WITH customer_sales AS (
    SELECT o.CustomerID, SUM(o.Total_order_amount) AS total_sales
    FROM orders o
    GROUP BY o.CustomerID
)
SELECT CustomerID, total_sales,
       RANK() OVER (ORDER BY total_sales DESC) AS sales_rank
FROM customer_sales
LIMIT 10;
""")


CustomerID  total_sales  sales_rank
0       57249     442545.0           1
1       57495     421414.0           2
2       57574     408974.0           3
3       57455     405883.0           4
4       57486     401145.0           5
5       57232     393376.0           6
6       57213     389207.0           7
7       57200     388967.0           8
8       57339     384503.0           9
9       57102     376974.0          10

## Insights  

- Customer base is **geographically diverse**, with the United States, India, and multiple European countries leading in volume. This indicates strong potential for global marketing strategies.  

- **Steady new customer acquisitions** from 2020–2021 reflect consistent growth, suggesting effective onboarding and awareness campaigns.  

- A small group of **high-frequency buyers** (e.g., Luca – 19 orders, Rex – 18) drive repeat sales. These customers are ideal for loyalty and retention programs.  

- **High-value customers** like Jacqueline Fernandez (₹455K+) and Toby Richard (₹425K+) contribute disproportionately to revenue. Retaining these key accounts is critical.  

- Analysis shows **frequency ≠ revenue**: some buy often with small baskets, while others purchase fewer but larger orders. Marketing should adapt to both patterns.  

- **Quick consecutive purchases** (within 7 days) highlight segments responding well to promotions or urgent demand. These insights can guide targeted re-marketing.  

- The **top 10 customers account for a large share of sales**, confirming that focusing on the top 20% can yield outsized impact on revenue and retention.  


---

## Products & Categories
#### This section evaluates the performance of products and categories, helping the company understand demand patterns, pricing effectiveness, and category-level trends. By identifying best-sellers, underpriced products, and products driving the highest revenue, management can make better decisions regarding pricing, promotions, and product assortment. It also highlights which categories are strong and where gaps exist in the catalog.

#### Product sale and market prices are displayed to compare pricing and detect discounts.

In [272]:
run_query("""
SELECT Product, Sale_Price, Market_price
FROM products;
""")

Product  Sale_Price  \
0            Original Disinfectant Toilet Cleaner Liquid         489   
1                             Surface Disinfectant Spray         257   
2      Harpic Disinfectant Toilet Cleaner Original200...          74   
3      Harpic Toilet Cleaner Liquid - Original 1 L + ...         462   
4           Disinfectant Bathroom Cleaner Liquid - Lemon         299   
...                                                  ...         ...   
22083                   Canned Sardine - in Tomato Sauce          75   
22084                                       Silver Belly          45   
22085                             Prawns - Pd Xtra Large         247   
22086   Ready to Cook - Large Prawns (Peeled & Deveined)         275   
22087                          Tuna Spread In Mayonnaise         185   

       Market_price  
0               534  
1               318  
2                76  
3               558  
4               362  
...             ...  
22083            79  
22084            50  
22085           290  
22086           305  
22087           195  

[22088 rows x 3 columns]

#### Active categories are listed to confirm which categories are currently offered.

In [273]:
run_query("""
SELECT CategoryName, Active
FROM category
WHERE Active = 'Yes';
""")

CategoryName Active
0      Cleaning & Household    Yes
1    Kitchen, Garden & Pets    Yes
2  Foodgrains, Oil & Masala    Yes
3      Gourmet & World Food    Yes
4    Snacks & Branded Foods    Yes
5     Bakery, Cakes & Dairy    Yes
6          Beauty & Hygiene    Yes
7                 Beverages    Yes
8       Fruits & Vegetables    Yes

#### Product counts in each sub-category are calculated to measure product diversity.

In [274]:
run_query("""
SELECT Sub_Category, COUNT(ProductID) AS cnt
FROM products
GROUP BY Sub_Category
ORDER BY cnt DESC;
""")

Sub_Category   cnt
0                     Skin Care  1988
1             Health & Medicine   975
2         Storage & Accessories   784
3              Bath & Hand Wash   753
4                     Hair Care   748
..                          ...   ...
85                    Non Dairy     5
86  Organic Fruits & Vegetables     5
87                    Marinades     1
88           Pork & Other Meats     1
89      Exotic Fruits & Veggies     1

[90 rows x 2 columns]

#### Products sold below market price are identified to track discounts and pricing opportunities.

In [275]:
run_query("""
SELECT ProductID, Product, Sale_Price, Market_Price
FROM products
WHERE Sale_Price < Market_Price;
""")

ProductID                                            Product  \
0              1        Original Disinfectant Toilet Cleaner Liquid   
1              2                         Surface Disinfectant Spray   
2              3  Harpic Disinfectant Toilet Cleaner Original200...   
3              4  Harpic Toilet Cleaner Liquid - Original 1 L + ...   
4              5       Disinfectant Bathroom Cleaner Liquid - Lemon   
...          ...                                                ...   
11710      22084                   Canned Sardine - in Tomato Sauce   
11711      22085                                       Silver Belly   
11712      22086                             Prawns - Pd Xtra Large   
11713      22087   Ready to Cook - Large Prawns (Peeled & Deveined)   
11714      22088                          Tuna Spread In Mayonnaise   

       Sale_Price  Market_Price  
0             489           534  
1             257           318  
2              74            76  
3             462           558  
4             299           362  
...           ...           ...  
11710          75            79  
11711          45            50  
11712         247           290  
11713         275           305  
11714         185           195  

[11715 rows x 4 columns]

#### Total quantities sold are calculated to highlight high-demand products.

In [276]:
run_query("""
SELECT Product, SUM(Quantity) AS total_quantity
FROM products AS p 
LEFT JOIN OrderDetails AS od 
    ON p.ProductID = od.ProductID
GROUP BY Product 
ORDER BY total_quantity DESC;
""")

Product  total_quantity
0                       Apple Cider Vinegar           151.0
1                     Extra Light Olive Oil           131.0
2              Turmeric Powder/Arisina Pudi           117.0
3              Durum Wheat Pasta - Farfalle           115.0
4                             Masala - Meat           114.0
...                                     ...             ...
21185                Olives - Black, Pitted             NaN
21186                     Jalapeno - Slices             NaN
21187  Olives - Stuffed With Pimiento Paste             NaN
21188                        Burger - Chips             NaN
21189                 Green Olives - Sliced             NaN

[21190 rows x 2 columns]

#### Sales revenue per product is computed to reveal top revenue contributors.

In [277]:
run_query("""
SELECT p.ProductID, p.Product, 
       SUM(od.Quantity * p.Sale_Price) AS total_sales
FROM products AS p 
JOIN orderdetails AS od 
    ON p.ProductID = od.ProductID 
GROUP BY p.ProductID, p.Product
ORDER BY total_sales DESC, p.Product ASC;
""")

ProductID                                            Product  \
0          18568      Stainless Steel Contura Pressure Cooker SSC50   
1          12576                    Raw Manuka Honey K Factor - 16+   
2           4711                              AQVA Divina Body Mist   
3           4817                       Voyage Sport Eau De Toilette   
4           4259                         Man In Black Eau De Parfum   
...          ...                                                ...   
15717      13835  Ultra-Fresh Fairness Face Wash With Milk, Lico...   
15718       3459                               Orange Chewy Dragees   
15719       3391                       Cadbury Perk - Chocolate Bar   
15720       2556                     Happy Happy Egg Cake - Vanilla   
15721       2095                          Marie Light Oats Biscuits   

       total_sales  
0         192864.0  
1         190800.0  
2         182000.0  
3         164736.0  
4         164700.0  
...            ...  
15717         13.0  
15718         11.0  
15719         10.0  
15720         10.0  
15721         10.0  

[15722 rows x 3 columns]

#### Products are linked with their categories to evaluate performance at both levels.

In [278]:
run_query("""
SELECT c.CategoryName, p.Product 
FROM Category AS c 
JOIN Products AS p 
    ON p.Category_ID = c.CategoryID 
ORDER BY c.CategoryName ASC;
""")

CategoryName  \
0                   Baby Care   
1                   Baby Care   
2                   Baby Care   
3                   Baby Care   
4                   Baby Care   
...                       ...   
22083  Snacks & Branded Foods   
22084  Snacks & Branded Foods   
22085  Snacks & Branded Foods   
22086  Snacks & Branded Foods   
22087  Snacks & Branded Foods   

                                                 Product  
0                                   Hippo Sipper - Green  
1                                     Sippie Cup - Green  
2      Baby Feeding Bottles For Milk & Water With Handle  
3                          Peristaltic Nipple - 'S' Hole  
4      Silicone Feeding Nipple/Teat Wide Neck Nipple ...  
...                                                  ...  
22083                                           Moti Pak  
22084      Organic Energy Paak - Sugar Free, Gluten Free  
22085              Sweet - Kheer Mohan, Brown Rassogolla  
22086                                    Chocolate Burfi  
22087                                        Besan Laddu  

[22088 rows x 2 columns]

#### The average sale price per category is calculated to analyze category pricing strategies.

In [279]:
run_query("""SELECT c.CategoryName, AVG(p.Sale_Price) AS average_sale_price
FROM Category AS c 
JOIN Products AS p 
    ON p.Category_ID = c.CategoryID 
GROUP BY c.CategoryName
ORDER BY c.CategoryName ASC;
""")

CategoryName  average_sale_price
0                  Baby Care            555.4139
1      Bakery, Cakes & Dairy            134.5545
2           Beauty & Hygiene            439.4021
3                  Beverages            244.6516
4       Cleaning & Household            223.9378
5          Eggs, Meat & Fish            285.2934
6   Foodgrains, Oil & Masala            167.8178
7        Fruits & Vegetables             64.9063
8       Gourmet & World Food            303.2058
9     Kitchen, Garden & Pets            520.6068
10    Snacks & Branded Foods            125.7925

#### The top 5 products by sales value are shown to identify best-sellers.

In [280]:
run_query("""
SELECT ProductID, Product, SUM(Sale_Price) AS total_sales_value
FROM products 
GROUP BY ProductID, Product
ORDER BY total_sales_value DESC
LIMIT 5;
""")

ProductID                                            Product  \
0      11012                                    Bravura Clipper   
1      10851      Pet Food - N&D Team Breeder Puppy Top Farmina   
2      21909                   Epilator SE9-9961 Legs-Body-Face   
3      18645  Gas Stove-4 Burner Royale Plus Schott Glass, B...   
4      18649  4 Burner Gas Stove - Marvel Plus Glass Tables,...   

   total_sales_value  
0            12500.0  
1            10090.0  
2             8184.0  
3             7999.0  
4             7270.0

#### Products are ranked by total revenue to highlight top-performing items.

In [281]:
run_query("""
SELECT p.ProductID, 
       SUM(p.Sale_Price * od.Quantity) AS total_revenue,
       DENSE_RANK() OVER(ORDER BY SUM(p.Sale_Price * od.Quantity) DESC) AS rnk 
FROM products AS p 
JOIN orderdetails AS od 
    ON p.ProductID = od.ProductID 
GROUP BY p.ProductID;
""")

ProductID  total_revenue   rnk
0          18568       192864.0     1
1          12576       190800.0     2
2           4711       182000.0     3
3           4817       164736.0     4
4           4259       164700.0     5
...          ...            ...   ...
15717      13835           13.0  5136
15718       3459           11.0  5137
15719       2556           10.0  5138
15720       3391           10.0  5138
15721       2095           10.0  5138

[15722 rows x 3 columns]

#### Cumulative revenue per product over time is calculated to observe sales progression.

In [282]:
run_query("""
SELECT p.Product, 
       o.OrderDate, 
       p.Sale_price * od.Quantity AS total_revenue, 
       SUM(p.Sale_price * od.Quantity) OVER(PARTITION BY p.Product ORDER BY o.OrderDate) AS cumulative_sum
FROM products AS p 
JOIN orderdetails AS od 
    ON p.ProductID = od.ProductID
JOIN orders AS o 
    ON od.OrderID = o.OrderID;
""",)

Product   OrderDate  \
0                                   & Moms - Citrus Soap  2021-07-01   
1                                   & Moms - Citrus Soap  2021-10-14   
2                          & Moms Bathing Soap - Jasmine  2021-11-08   
3                            & Moms Bathing Soap - Tulsi  2020-10-10   
4                            & Moms Bathing Soap - Tulsi  2021-08-04   
...                                                  ...         ...   
27522                             Zipper Soulful Jasmine  2021-06-03   
27523                            Zipper Storage - Gallon  2021-06-24   
27524  Zookers Animal Shaped Biscuits - Cherry Bits, ...  2021-04-09   
27525                Zoom Ultimate Dark Triangle Pencils  2020-10-04   
27526                Zoom Ultimate Dark Triangle Pencils  2021-05-14   

       total_revenue  cumulative_sum  
0               2730          2730.0  
1                195          2925.0  
2                972           972.0  
3               2160          2160.0  
4                216          2376.0  
...              ...             ...  
27522           1100          1430.0  
27523           3150          3150.0  
27524           2240          2240.0  
27525            735           735.0  
27526           2058          2793.0  

[27527 rows x 4 columns]

#### Discounts are measured by comparing market and sale price to rank products with the highest markdowns.

In [283]:
run_query("""
SELECT Product, 
       Market_Price, 
       Sale_Price, 
       Market_Price - Sale_Price AS price_difference, 
       DENSE_RANK() OVER(ORDER BY (Market_Price - Sale_Price) DESC) AS rank_
FROM products;
""")

Product  Market_Price  \
0      Premium Cloth Dryer/Drying Stand - Foldable, S...          8969   
1      Gas Stove-4 Burner Royale Plus Schott Glass, B...         12245   
2      Gas Stove-3 Burner Royale Plus Schott Glass, B...         10695   
3      Induction Base Non Stick Cookware Set With Gla...          4999   
4      Induction Base Non Stick Cookware Set With Gla...          4999   
...                                                  ...           ...   
22083                             Prostate Care Capsules           225   
22084     Kidney Detox With Varun, Punerneva & Ganoderma          1499   
22085  Instantized BCAA Enriched with Electrolytes - ...          1500   
22086                           Ayurvedic Satawar Powder           120   
22087                                Mahadrakshadi Syrup            90   

       Sale_Price  price_difference  rank_  
0            4649              4320      1  
1            7999              4246      2  
2            6999              3696      3  
3            1399              3600      4  
4            1499              3500      5  
...           ...               ...    ...  
22083         225                 0    684  
22084        1499                 0    684  
22085        1500                 0    684  
22086         120                 0    684  
22087          90                 0    684  

[22088 rows x 5 columns]

#### The best-selling product in each category is identified to reveal category leaders.

In [284]:
run_query("""
WITH top_selling_product AS (
    SELECT c.CategoryID, 
           c.CategoryName, 
           Product, 
           SUM(od.Quantity) AS total_quantity, 
           DENSE_RANK() OVER(PARTITION BY c.CategoryID ORDER BY SUM(od.Quantity) DESC) AS rank_
    FROM category AS c 
    JOIN products AS p 
        ON c.CategoryID = p.Category_ID 
    JOIN orderdetails AS od 
        ON p.ProductID = od.ProductID
    GROUP BY c.CategoryID, c.CategoryName, Product
)
SELECT CategoryName, Product, total_quantity
FROM top_selling_product
WHERE rank_ = 1 
ORDER BY total_quantity DESC;
""")

CategoryName                                   Product  \
0   Foodgrains, Oil & Masala              Turmeric Powder/Arisina Pudi   
1       Gourmet & World Food              Durum Wheat Pasta - Farfalle   
2       Cleaning & Household                            Toilet Cleaner   
3     Kitchen, Garden & Pets    Plain Container Jumbo Combo Set - Pink   
4      Bakery, Cakes & Dairy                            Cheese - Block   
5           Beauty & Hygiene                                Castor Oil   
6     Snacks & Branded Foods                 Strawberry Creme Biscuits   
7                  Beverages                      Chai - Masala, Tulsi   
8                  Baby Care     Reusable Fabric Diaper - Medium, Blue   
9          Eggs, Meat & Fish  Tuna Salad With Vegetables Mexican Style   
10       Fruits & Vegetables               Sambhar Onion (Small Onion)   

    total_quantity  
0            117.0  
1            115.0  
2             94.0  
3             88.0  
4             82.0  
5             82.0  
6             75.0  
7             69.0  
8             65.0  
9             54.0  
10            54.0

## Insights  

- The **product catalog is extensive and diverse**, spanning essentials like cleaning, food, and personal care to premium categories such as gourmet foods and luxury beauty products.  

- Active categories cover all major customer needs, with strong representation across **daily essentials (Foodgrains, Beverages, Cleaning, Bakery, Dairy)** and **lifestyle products (Beauty, Kitchen, Baby Care)**. This wide spread helps in capturing multiple customer segments.  

- Sub-categories such as **Skin Care (1,988 products)** and **Health & Medicine (975 products)** dominate, showing significant depth in personal care and wellness—indicating where customers may be offered bundles and targeted campaigns.  

- A large number of products are sold **below market price**, highlighting the company’s competitive pricing strategy to attract price-sensitive customers. At the same time, significant **discount gaps** (e.g., ₹4,000+ on premium appliances) suggest opportunities to market value-for-money positioning.  

- **Top-selling items** like Apple Cider Vinegar, Olive Oil, and Spices indicate strong traction in health-conscious and cooking essentials categories, aligning with rising consumer trends.  

- High-revenue drivers such as **Pressure Cookers, Premium Honey, and Branded Perfumes** demonstrate that premium household and lifestyle products are key contributors to profitability.  

- Average sale prices show clear segmentation: **Baby Care (~₹555)** and **Kitchen products (~₹520)** skew higher, while **Fruits & Vegetables (~₹65)** and **Snacks (~₹125)** remain low-ticket but volume-driven.  

- Revenue distribution reveals a **long-tail effect**: while a few products generate very high sales, many products contribute modestly. This suggests cross-selling and recommendation systems could unlock more value from underperforming products.  

- Cumulative sales analysis confirms steady **repeat demand across time** for daily essentials, while premium products show **spikes during promotions or seasonal events**.  

- Ranking products by discount percentage highlights strong **value-proposition opportunities** in home and kitchen appliances, which can be leveraged in festive and sale campaigns.  

- At the category level, **Foodgrains, Household, and Gourmet Foods** consistently appear in top sellers, making them pillars of the business. This reinforces the need to optimize supply chains and ensure high availability in these segments.  


---

## Orders & Revenue Trends
#### This section focuses on order management and revenue trends over time. By analyzing order timelines and monthly sales, management can measure operational efficiency and detect seasonal demand patterns. These insights are critical for planning promotions, managing inventory, and improving delivery timelines to meet customer expectations.

#### Orders with order and ship dates are displayed to assess fulfillment and delivery cycles.

In [285]:
run_query("""
SELECT OrderID, OrderDate, ShipDate
FROM orders;
""")

OrderID   OrderDate    ShipDate
0     7655500  2020-01-12  2020-01-13
1     7655501  2020-01-20  2020-01-24
2     7655502  2020-02-06  2020-02-11
3     7655503  2020-02-09  2020-02-13
4     7655504  2020-02-11  2020-02-15
...       ...         ...         ...
4994  7660494  2021-12-28  2021-12-29
4995  7660495  2021-12-28  2021-12-29
4996  7660496  2021-12-28  2022-01-04
4997  7660497  2021-12-28  2022-01-01
4998  7660498  2021-12-28  2022-01-05

[4999 rows x 3 columns]

#### Monthly revenue trends are calculated to uncover seasonal sales fluctuations and high-demand periods

In [286]:
run_query("""
SELECT MONTH(OrderDate) AS month_, 
       MONTHNAME(OrderDate) AS month_name, 
       SUM(od.Quantity * p.Sale_Price) AS total_revenue 
FROM orders AS o 
JOIN OrderDetails AS od 
    ON o.OrderID = od.OrderID 
JOIN Products AS p 
    ON od.ProductID = p.ProductID 
GROUP BY month_, month_name 
ORDER BY month_ ASC;
""")

month_ month_name  total_revenue
0        1    January      2811699.0
1        2   February      2886049.0
2        3      March      4649677.0
3        4      April      4980880.0
4        5        May      5979434.0
5        6       June      6492098.0
6        7       July      8126607.0
7        8     August      8193161.0
8        9  September      9314621.0
9       10    October     11653090.0
10      11   November     12992401.0
11      12   December     15415899.0

## Insights  

- Order activity is **consistent across 2020–2021**, with a stable flow of shipments being processed almost daily. The short gap between **order date and ship date** highlights efficient operational handling.  

- Monthly revenue analysis shows a **clear upward trend** through the year, with sales peaking in **October, November, and December**. This aligns with festive and holiday shopping seasons, confirming the importance of targeted campaigns during these months.  

- The company’s **highest revenue month is December (~₹15.4M)**, nearly **5x the revenue of January**, showing strong seasonality and the impact of year-end promotions.  

- Mid-year (June–August) also records solid growth (~₹8M+), possibly linked to seasonal demand or mid-year sales, indicating additional opportunities for promotional pushes outside the festive season.  

- Early months (January–April) contribute relatively lower sales, suggesting that **marketing efforts and engagement strategies** in Q1 could be improved to balance yearly revenue distribution.  

- Overall, the trend demonstrates a **strong festive-driven business model** with opportunities to drive more stable revenue across off-peak months through loyalty programs, cross-selling, and subscription models.  


---

## Suppliers & Shipping
#### This section assesses supplier contributions and logistics efficiency, which are central to maintaining product availability and customer satisfaction. By tracking which suppliers contribute the most, how revenues compare across suppliers, and how quickly shippers deliver, management can optimize supplier relationships and streamline logistics operations.

#### Suppliers offering more than five products are identified to highlight key partners with broad catalogs.

In [287]:
run_query("""
SELECT s.SupplierID, s.CompanyName, COUNT(DISTINCT p.ProductID) AS cnt
FROM Suppliers AS s 
JOIN OrderDetails AS od ON s.SupplierID = od.SupplierID 
JOIN Products AS p ON od.ProductID = p.ProductID 
GROUP BY s.SupplierID, s.CompanyName
HAVING COUNT(DISTINCT p.Product) > 5
ORDER BY cnt DESC;
""")

SupplierID                     CompanyName   cnt
0           4                    GreenDroShip  4158
1           6                    ASI Partners  4155
2           3                   Alpha imports  4138
3           2                   Tasha Apparel  4128
4           1  Sunshine Electronics Pvt. Ltd.  4113
5           5                       Megagoods  4112

#### Supplier revenues are calculated and compared with the average to identify high-performing suppliers.

In [288]:
run_query("""
WITH supplier_revenue AS (
    SELECT s.SupplierID, s.CompanyName, 
           ROUND(SUM(od.Quantity * p.Sale_Price), 2) AS total_revenue
    FROM suppliers AS s 
    JOIN orderdetails AS od ON s.SupplierID = od.SupplierID 
    JOIN products AS p ON od.ProductID = p.ProductID
    GROUP BY s.SupplierID, s.CompanyName
),
overall_revenue AS (
    SELECT ROUND(AVG(total_revenue), 0) AS average_revenue
    FROM supplier_revenue
)
SELECT SupplierID, CompanyName, total_revenue, average_revenue 
FROM supplier_revenue, overall_revenue 
WHERE total_revenue > average_revenue;
""")

SupplierID    CompanyName  total_revenue  average_revenue
0           2  Tasha Apparel     15659519.0       15582603.0
1           4   GreenDroShip     16087094.0       15582603.0
2           6   ASI Partners     16277898.0       15582603.0

#### Orders shipped by each shipping partner are counted to evaluate workload distribution.

In [289]:
run_query("""
SELECT s.ShipperID, s.CompanyName, COUNT(o.OrderID) AS cnt 
FROM shippers AS s 
JOIN orders AS o ON o.ShipperID = s.ShipperID 
GROUP BY s.ShipperID, s.CompanyName
ORDER BY cnt DESC, s.CompanyName ASC;
""")

ShipperID                           CompanyName  cnt
0          8           ONE - Ocean Network Express  691
1          3                       Lufthansa Cargo  641
2          1                             Blue Dart  625
3          5                             Delhivery  620
4          2                                  DTDC  613
5          6  COSCO - China Ocean Shipping Company  611
6          7                           Hapag-Lloyd  601
7          4                                Fed Ex  597

#### Average delivery days per shipper are measured to find the fastest and most reliable logistics provider

In [290]:
run_query("""
SELECT s.ShipperID, s.CompanyName,
       ROUND(AVG(TIMESTAMPDIFF(DAY, o.OrderDate, o.DeliveryDate)),2) AS average_delivery_days
FROM shippers AS s 
JOIN orders AS o ON s.ShipperID = o.ShipperID 
GROUP BY s.ShipperID, s.CompanyName
ORDER BY average_delivery_days ASC
LIMIT 1;
""")

ShipperID                  CompanyName  average_delivery_days
0          8  ONE - Ocean Network Express                  15.67

#### Shipping Efficiency (Delivery Performance)

In [291]:
run_query("""
SELECT s.CompanyName AS shipper_name,
       AVG(DATEDIFF(o.DeliveryDate, o.OrderDate)) AS avg_delivery_days,
       MIN(DATEDIFF(o.DeliveryDate, o.OrderDate)) AS fastest_delivery,
       MAX(DATEDIFF(o.DeliveryDate, o.OrderDate)) AS slowest_delivery
FROM orders o
JOIN shippers s ON o.ShipperID = s.ShipperID
GROUP BY s.CompanyName
ORDER BY avg_delivery_days;
""")

shipper_name  avg_delivery_days  fastest_delivery  \
0           ONE - Ocean Network Express            15.6744                 2   
1  COSCO - China Ocean Shipping Company            15.8658                 2   
2                             Delhivery            16.0806                 2   
3                                Fed Ex            16.1457                 2   
4                             Blue Dart            16.1984                 2   
5                       Lufthansa Cargo            16.2044                 2   
6                                  DTDC            16.2610                 2   
7                           Hapag-Lloyd            16.2679                 2   

   slowest_delivery  
0                30  
1                30  
2                30  
3                30  
4                30  
5                30  
6                30  
7                30

## Insights  

- Supplier contribution is well-distributed across a few key players, with **GreenDroShip, ASI Partners, and Alpha Imports** handling the majority of supply volumes (~4,100+ each). This indicates strong supplier partnerships and reduced dependency risk.  

- In terms of revenue, **ASI Partners leads (~₹16.2M)**, closely followed by GreenDroShip and Tasha Apparel, showing that supplier performance is **balanced and competitive**. Maintaining strong relationships with these top suppliers is critical for stable revenue flow.  

- On the logistics side, **ONE - Ocean Network Express and Lufthansa Cargo** are the most frequently used shippers, processing ~690 and ~640 shipments respectively, showing reliability and preference in operations.  

- Delivery performance across shippers is fairly similar, averaging **15–16 days**, with fastest deliveries achieved in **2 days** and slowest extending to **30 days**. This consistency indicates predictable service but also highlights scope for improvement in reducing overall lead times.  

- **Delhivery and Blue Dart** show slightly better average delivery times compared to global freight providers, making them key partners for **domestic efficiency**.  

- Overall, suppliers are performing strongly, and shipping networks are stable, but reducing the **delivery window (15–16 days)** could significantly improve customer satisfaction and retention.  


---

# Final Business Recommendation  

The e-commerce company has a strong foundation — a **global customer base**, **broad product catalog**, **balanced suppliers**, and **stable shipping partners**. But profitability and retention are limited by **deep discounts, seasonal spikes, long delivery lead times (~15–16 days)**, and a **long tail of low-performing SKUs**.  

To unlock growth, the company should integrate **customer, product, revenue, supplier, and logistics strategies** into one cohesive plan:  


## Customer Growth & Retention  
- Focus on **top 20% high-value customers** with **VIP loyalty programs, premium memberships, and early access deals**.  
- Segment marketing:  
  - **Frequent low-basket buyers** → subscriptions, curated bundles, rewards to lift AOV.  
  - **High-value but infrequent buyers** → exclusivity, concierge-like service to retain spend.  
- Use **behavioral triggers** (repeat purchases within 7 days) for personalized replenishment offers.  


## Product & Pricing Optimization  
- Protect **daily essentials** (Foodgrains, Household, Gourmet Foods) as reliable revenue anchors.  
- Scale **wellness & premium lifestyle products** (Apple Cider Vinegar, Olive Oil, Skincare, Perfumes).  
- **Prune unproductive SKUs** and surface the long tail with **cross-selling & recommendation engines**.  
- Shift to **margin-aware discounting**: use supplier-funded promos for appliances, avoid site-wide deep cuts.  


## Revenue & Seasonality  
- Smooth the **festive-heavy cycle (Dec = 5× Jan)**:  
  - Boost **Q1** with wellness bundles, household savings packs, and subscriptions.  
  - Scale **mid-year campaigns** with seasonal hooks (monsoon / Independence Day).  
- Push **subscriptions for essentials** to stabilize revenue year-round.  


## Supplier & Logistics Excellence  
- Deepen partnerships with **ASI Partners, GreenDroShip, Tasha Apparel** for co-funded promos and inventory assurance.  
- Reduce delivery lead time by:  
  - Prioritizing **fast domestic shippers (Delhivery, Blue Dart)**.  
  - Building **regional fulfillment hubs** to pre-stock high-demand SKUs.  
  - Using **predictive forecasting** to cut delays during peaks.  
- Target reducing delivery from **15–16 days → under 7 days**.  





---

# Key Takeaways  

- **Top customers drive revenue** → invest in loyalty, exclusivity, and retention.  
- **Frequent small-basket buyers ≠ high spenders** → use bundles, subscriptions, and upselling.  
- **Essentials + premium lifestyle products are core pillars** → ensure strong availability and supply chain focus.  
- **Discounts must be margin-aware** → shift from blanket cuts to targeted, supplier-supported promotions.  
- **Revenue is overly festive-driven** → balance seasonality with Q1 and mid-year campaigns.  
- **Suppliers are reliable but logistics lag** → reduce delivery times from 15–16 days to under 7 with regional hubs and faster domestic carriers.  
 

---